graphql-ballerina/src/graphql/parser/lexer.bal

```diff
@@ -0,0 +1,325 @@
+    private CharReader charReader;
+    private string document;
+    private boolean inProgress;
+    private Token[] buffer;
+
+    public isolated function init(string document) {
+        self.charReader = new(document);
+        self.document = document;
+        self.inProgress = true;
+        self.buffer = [];
+    }
+
+    public isolated function reset() {
+        self.charReader = new(self.document);
+        self.buffer = [];
+    }
+
+    isolated function peekLexical() returns Token|ParsingError {
+        Token? token = ();
+        int i = 1;
+        while (true) {
+            Token nextToken = check self.peek(i);
+            if (nextToken.'type is LexicalType) {
+                token = nextToken;
+                break;
+            }
+            i += 1;
+        }
+        return <Token>token;
+    }
+
+    public isolated function peek(int n = 1) returns Token|ParsingError {
+        int index = n - 1;
+        if (self.buffer.length() < n) {
+            int difference = n - self.buffer.length();
+            int i = 0;
+            while (i <= difference) {
+                Token token = check self.readNextToken();
+                self.pushToBuffer(token);
+                i += 1;
+            }
+        }
+        return self.buffer[index];
+    }
+
+    public isolated function hasNext() returns boolean {
+        if (self.buffer.length() > 0) {
+            return true;
+        }
+        return self.inProgress;
+    }
+
+    isolated function nextLexicalToken() returns Token|ParsingError {
+        Token? token = ();
+        while (true) {
+            Token nextToken = check self.next();
+            if (nextToken.'type is LexicalType) {
+                token = nextToken;
+                break;
+            }
+        }
+        return <Token>token;
+    }
+
+    public isolated function next() returns Token|ParsingError {
+        if (self.buffer.length() > 0) {
+            return self.buffer.shift();
+        }
+        return self.readNextToken();
+    }
+
+    isolated function readNextToken() returns Token|ParsingError {
+        Char char = self.charReader.next();
+        TokenType tokenType = getTokenType(char);
+        if (tokenType == T_EOF) {
+            self.inProgress = false;
+            return getTokenFromChar(char);
+        } else if (tokenType == T_STRING) {
+            return self.readStringToken(char.location);
+        } else if (tokenType == T_INT) {
+            return self.readNumeralToken(char.location, char.value);
+        } else if (tokenType is TerminalCharacter) {
+            return getTokenFromChar(char);
+        } else if (tokenType == T_COMMENT) {
+            return self.readCommentToken(char.location);
+        } else if (tokenType is SpecialCharacter) {
+            return getTokenFromChar(char);
+        } else {
+            return self.readWordToken(char);
+        }
+    }
+
+    isolated function getNextSpecialCharaterToken() returns Token|ParsingError {
```

Comment:
even though this method is called `getNextSpecialCharaterToken`, looks like it returns the next non-specical character token?